### PR TITLE
Extend the distributed implementation to support Sumpy

### PR DESCRIPTION
**Pull request by gaohao95**
- Make the `DistributedFMMRunner` accept trees both in the host and the device memory.
- Use structure of arrays instead of multi-dimentional arrays for particles passed into kernels.

After merge:
- Look at https://github.com/inducer/sumpy/pull/137

### CODE REVIEW

File: boxtree/distributed/__init__.py [hunk 1]
```diff
@@ -160,16 +164,37 @@ def __init__(self, queue, global_tree_dev,
         self.comm = comm
         mpi_rank = comm.Get_rank()
 
+        tree_in_device_memory = None
```
Comment (inducer) [inline boxtree/distributed/__init__.py:167]: Explain the purpose of this flag.

File: boxtree/distributed/__init__.py [hunk 2]
```diff
@@ -138,13 +140,15 @@ class DistributedFMMRunner:
     .. automethod:: __init__
     .. automethod:: drive_dfmm
     """
-    def __init__(self, queue, global_tree_dev,
+    def __init__(self, queue, global_tree,
```
Comment (inducer) [inline boxtree/distributed/__init__.py:143]: Make a "maker" function, and have the constructor do no work.

**Review by inducer — APPROVED** [review]
Thanks! LGTM after these two small things are fixed.

**Comment by inducer** [comment]
This has picked up some conflicts. Could you take a look?

**Comment by gaohao95** [comment]
> This has picked up some conflicts. Could you take a look?

Rebased (and squashed).

**Comment by inducer** [comment]
Thx!

**Pull request merged**